### PR TITLE
Validation single subscription field

### PR DESCRIPTION
**Pull request by faizan-siddiqui**
Add validation for single root subscription field only

### CODE REVIEW

File: src/test/groovy/graphql/validation/rules/SingleSubscriptionFieldOnlyTest.groovy [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import graphql.validation.ValidationError
+import graphql.validation.ValidationErrorType
+import graphql.validation.Validator
+import spock.lang.Specification
+
+import static graphql.validation.rules.SingleSubscriptionFieldOnly.moreThanOneSubscriptionMessage
+
+class SingleSubscriptionFieldOnlyTest extends Specification {
+
+    def '5.2.3.1 Subscription operations must have exactly one root field.'() {
```
Comment (andimarek) [inline src/test/groovy/graphql/validation/rules/SingleSubscriptionFieldOnlyTest.groovy:15]: can we remove the spec section references? It has not really payed of over the long time and it is also not guaranteed to be stable.

File: src/main/java/graphql/validation/rules/SingleSubscriptionFieldOnly.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public void checkOperationDefinition(OperationDefinition operationDefinition) {
+        super.checkOperationDefinition(operationDefinition);
+
+        // skip validation for non subscription operations
+        if (!isSubscriptionOperation(operationDefinition)) {
+            return;
+        }
+
+        List<Selection> selections = operationDefinition.getSelectionSet().getSelections();
+        if (selections.size() > 1) {
```
Comment (bbakerman) [inline src/main/java/graphql/validation/rules/SingleSubscriptionFieldOnly.java:38]: Close but not quite.

This needs to be the output of CollectFields - eg `graphql.execution.FieldCollector`

However I am not sure we can do that in validation rules can we??

I checked and there is no code using `graphql.execution.FieldCollector` today in validation

Is it ok to do so?

If you look at the spec there give and example of collected fields

Comment (bbakerman) [reply src/main/java/graphql/validation/rules/SingleSubscriptionFieldOnly.java:38]: Single root field

Formal Specification

For each subscription operation definition subscription in the document
* Let subscriptionType be the root Subscription type in schema.
* Let selectionSet be the top level selection set on subscription.
* Let variableValues be the empty set.
* Let groupedFieldSet be the result of `CollectFields(subscriptionType, selectionSet, variableValues)`.
* groupedFieldSet must have exactly one entry.

Comment (faizan-siddiqui) [reply src/main/java/graphql/validation/rules/SingleSubscriptionFieldOnly.java:38]: the FieldCollector methods require a lot more parameters than what is available in validation, which just traverses through the nodes, will have to see if it is even possible to provide the parameters, when in the context of validation we only have the Node object, and not other things like schema

Comment (bbakerman) [reply src/main/java/graphql/validation/rules/SingleSubscriptionFieldOnly.java:37]: See `graphql.validation.ValidationContext` - every validation rule has it.  If we need more (I think we need variables) then we should extend it to contain what we need

The schema and fragments are all present

**Comment by andimarek** [comment]
Please have a look at https://github.com/graphql/graphql-spec/pull/860 and https://github.com/graphql/graphql-spec/pull/806 too.

**Pull request closed without merging**